### PR TITLE
feat: base game entity, state machines and messaging system

### PR DESCRIPTION
**Pull request by benekoehler**

### CODE REVIEW

File: Entity/Enemy/Enemy.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+// Created by benek on 10/14/24.
+//
+
+#include "Enemy.h"
+#include "EnemyStates.h"
+#include <iostream>
+
+void Enemy::Update() {
+	m_pStateMachine->Update();
+	std::cout << "Enemy Health: " << m_Health << "\n";
```
Comment (benekoehler) [inline Entity/Enemy/Enemy.cpp:11]: i currently write a lot of stuff to cout. this is not intended to land in prodcution just for easier debugging until we marry the rendering of actual models with the entities and stuff

File: Entity/Player/Player.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+	if (KeyPressed(SDLK_s)) {
+		m_pStateMachine->ChangeState(PlayerIdle::Instance());
+	}
+
+	if (KeyPressed(SDLK_SPACE)) {
+		m_pStateMachine->ChangeState(PlayerAttacking::Instance());
+	}
+	m_pStateMachine->Update();
+}
```
Comment (benekoehler) [inline Entity/Player/Player.cpp:25]: i am curious what you have in mind about input handling? is this okeyish?

File: Entity/Player/Player.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	bool HandleMessage(const Telegram &telegram) override;
+
+	bool CanAttack() {
+		double currentTime = Clock->GetTime();
+		if (currentTime >= m_LastAttack + m_AttackDelay) {
+			m_LastAttack = currentTime;
+			return true;
+		}
+		return false;
+	}
```
Comment (benekoehler) [inline Entity/Player/Player.h:35]: just as an example. this should be moved to the weapon when we get have it mapped out :)

File: Message/MessageDispatcher.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	// create the telegram
+	Telegram telegram(0, sender, receiver, message, ExtraInfo);
+
+	// if there is no delay, route telegram immediately
+	if (delay <= 0.0f) {
+		std::cout << "\nInstant telegram dispatched at time: " << Clock->GetTime() << " by " << pSender->ID() << " for "
+				  << pReceiver->ID() << ". Msg is " << MessageToString(message) << "\n";
+		Discharge(pReceiver, telegram);
+	} else {
+		// TODO add the delay
```
Comment (benekoehler) [inline Message/MessageDispatcher.cpp:56]: cleanup unused comment

File: Entity/Enemy/Enemy.cpp (hunk 1 above)
Comment (michaeleggers) [reply Entity/Enemy/Enemy.cpp:11]: No problem writing to stdout for now. But don't use iostream for that. See tech. design document (orthodox c++). I also use iostreams in the map parser. It is a wrongdoing I did in the past and needs to be fixed.

File: Entity/Player/Player.cpp (hunk 2 above)
Comment (michaeleggers) [reply Entity/Player/Player.cpp:25]: Yup. For now this is fine!

**Comment by michaeleggers** [comment]
Overall, this looks good!

Just get rid of `cout` and overloading the `<<` operator for printing. Also don't get in the STL's version of C headers. There are some modern C++ things in it like iterating over the map via auto. I think this is fine for now. Just note that we are going to replace the entire STL later. Look at id-tech3 source, as mentioned before.
Giving this more tests later...

**Comment by michaeleggers** [comment]
Thank you for the changes.
When including stuff from the C standard library, use `<... .h>` instad of  `"... .h"`. Also include `stdio.h` where `printf` is used.

**Comment by michaeleggers** [comment]
Okay. We're getting into a very important stage of the program now: Looking how we can integrate entities and their logic without creating a really big mess. @benekoehler is going to tackle this (good luck ;) ).
A few notes that I want to see (most of it was already communicated in person with @benekoehler):

- Think in systems and data, eg: Do not let the entities render themselves. The renderer just takes the entities (or maybe even just their models) and draws them. As of now I would not change anything regarding the renderer. Just give the renderer the entities' model data, just like it is being done in `game.cpp`!
- Try not to have pointers to the world within the entities. Again: think in larger systems: If possible implement some logic in `CWorld` rather than have entities do things by themselves. Take this advice with a little bit caution: **If it is the easiest thing to get stuff done and is for everyone the easiest code to understand than storing a pointer to CWorld might be okay (for now)**.
- If you (@benekoehler ) get into trouble to write *clean code*, eg. you need globals that go over compilation units bounds, do other things that are *bad practice* according to frauds and bozos like Robert C. Martin (Uncle Bob), then just ***DO IT***. The goal is to get the steering behaviour running that you have researched. It is **always** better to create as little of abstractions as possible so the fewer classes we introduce the better. 
- Code that can be easily deleted without having to think about 1000 other things is good code! Eg: The deletion of a global variable that other systems rely on will break the whole program but it is **trivial** to fix! You just go into the files and search replace it with something else (a constant value until you have thought of something better, etc.). Whereas the instances of a Class that hold references to another class that executes logic every frame and does other things, because one thaught "it looks nice, everything is so tightly encapsulated...".... Good luck with that. It might not be bad at the beginning but the further one goes into OOP land the harder it gets to delete code. I don't reject OOP per se but one should be ***SUPER CAREFUL***. What we were taught at university about how to write software is largelay a pile of garbage (sorry to all the professors who taught me 'Software Engineering', but also not sorry! It just sucks.).
- Don't pay too much attention to performance, eg. there is a big fix comment in game.cpp that copying the brush entities' geometry into the world geometry vector *every frame* to make the collision system happy is super dumb. However, we are lucky to have fast CPUs and it does the job for now. We cannot fix everything at the same time! No problem doing this for now as it is trivially deletable (see above). 

I care  *a lot* about this. At the same time I am guilty of all of the things I mentioned above: I tend to overcomplicate things, and it *always* has hurt me in the past. I know this is not easy to get right and I don't have the absolute answer to how entities should work. I am looking forward to seee what you come up with @benekoehler . Good luck :100: 

Cheers.

**Review by michaeleggers — APPROVED** [review]

**Pull request merged**